File: android-kotlin/app/src/main/java/com/example/zipquantum/ZipQuantumClient.kt
package com.example.zipquantum

import android.content.Context
import android.net.Uri
import android.os.Build
import android.util.DisplayMetrics
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale
import java.util.TimeZone

class ZipQuantumClient(private val context: Context) {
    suspend fun resolve(uri: Uri): Delivery {
        require(uri.scheme == "https" && uri.host in ZQConfiguration.ALLOWED_HOSTS) { "Unconfigured App Link" }
        val reference = uri.lastPathSegment?.takeIf(String::isNotBlank) ?: error("Missing reference")
        return post("/api/mobile/v1/links/resolve", baseContext().apply {
            put("host", uri.host)
            put("reference", reference)
            put("url", uri.toString())
            put("package_name", ZQConfiguration.PACKAGE_NAME)
            put("parameters", JSONObject().also { target -> uri.queryParameterNames.forEach { target.put(it, uri.getQueryParameter(it)) } })
        })
    }

    suspend fun recover(handoff: DeferredHandoff): Delivery =
        post("/api/mobile/v1/deferred/recover", baseContext().apply {
            put("token", handoff.token)
            put("host", handoff.host)
            put("package_name", ZQConfiguration.PACKAGE_NAME)
        })

    suspend fun acknowledgeRouteOpened(ack: RouteAcknowledgement, host: String) {
        val endpoint = Uri.parse(ack.endpoint)
        val url = if (endpoint.isAbsolute) endpoint else Uri.parse(ZQConfiguration.API_BASE_URL).buildUpon().encodedPath(ack.endpoint).build()
        require(url.scheme == "https" && url.host == Uri.parse(ZQConfiguration.API_BASE_URL).host) { "Untrusted acknowledgement endpoint" }
        postObject(url.toString(), JSONObject().apply {
            put("receipt", ack.receipt)
            put("host", host)
            put("platform", "Android")
            put("package_name", ZQConfiguration.PACKAGE_NAME)
        })
    }

    private fun baseContext(): JSONObject {
        val metrics: DisplayMetrics = context.resources.displayMetrics
        return JSONObject().apply {
            put("platform", "Android")
            put("os_name", "Android")
            put("os_version", Build.VERSION.RELEASE)
            put("model", Build.MODEL)
            put("language", Locale.getDefault().toLanguageTag())
            put("timezone", TimeZone.getDefault().id)
            put("screen_resolution", "${metrics.widthPixels}x${metrics.heightPixels}")
            put("hardware_concurrency", Runtime.getRuntime().availableProcessors())
            put("tracking_consent", false)
            put("consent_version", ZQConfiguration.CONSENT_VERSION)
        }
    }

    private suspend fun post(path: String, body: JSONObject): Delivery {
        val response = postObject(ZQConfiguration.API_BASE_URL + path, body)
        if (!response.optBoolean("success")) error("Unsuccessful delivery")
        val link = response.getJSONObject("link")
        val ack = response.optJSONObject("route_ack")?.let {
            RouteAcknowledgement(it.getString("receipt"), it.getString("endpoint"), it.getInt("expires_in"))
        }
        return Delivery(
            response.getString("delivery"),
            ResolvedLink(link.getString("url"), link.optString("host").ifBlank { null }, link.optString("destination_url").ifBlank { null }),
            ack
        )
    }

    private suspend fun postObject(url: String, body: JSONObject): JSONObject = withContext(Dispatchers.IO) {
        val connection = URL(url).openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.connectTimeout = 12_000
            connection.readTimeout = 15_000
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.setRequestProperty("Accept", "application/json")
            connection.outputStream.use { it.write(body.toString().toByteArray()) }
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val text = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
            if (status !in 200..299) error("ZipQuantum HTTP $status")
            if (text.isBlank()) JSONObject() else JSONObject(text)
        } finally { connection.disconnect() }
    }
}

File: react-native/src/ZipQuantumClient.ts
import {Dimensions, Platform} from 'react-native';
import {zqConfig} from './config';
import {parseDelivery} from './delivery';
import {trustedAcknowledgementURL} from './handoff';
import type {DeferredHandoff, Delivery, RouteAcknowledgement} from './types';

const platform = Platform.OS === 'ios' ? 'iOS' : 'Android';

function baseContext(): Record<string, unknown> {
  const screen = Dimensions.get('screen');
  return {
    platform,
    os_name: platform,
    os_version: String(Platform.Version),
    language: Intl.DateTimeFormat().resolvedOptions().locale,
    timezone: Intl.DateTimeFormat().resolvedOptions().timeZone,
    screen_resolution: `${Math.round(screen.width * screen.scale)}x${Math.round(screen.height * screen.scale)}`,
    tracking_consent: false,
    consent_version: zqConfig.consentVersion,
  };
}

async function postJSON<T>(url: string, body: Record<string, unknown>): Promise<T> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 15_000);
  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {'Content-Type': 'application/json', Accept: 'application/json'},
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    if (!response.ok) throw new Error(`ZipQuantum HTTP ${response.status}`);
    const text = await response.text();
    return (text ? JSON.parse(text) : {}) as T;
  } finally {
    clearTimeout(timeout);
  }
}

function appBinding(): Record<string, string> {
  return Platform.OS === 'ios'
    ? {bundle_id: zqConfig.appIdentifier}
    : {package_name: zqConfig.appIdentifier};
}

export async function resolveDirectLink(value: string): Promise<Delivery> {
  const url = new URL(value);
  if (url.protocol !== 'https:' || !zqConfig.allowedHosts.some(host => host === url.host)) {
    throw new Error('Unconfigured Universal Link or App Link');
  }
  const reference = url.pathname.split('/').filter(Boolean).at(-1);
  if (!reference) throw new Error('Missing link reference');
  const response = await postJSON<unknown>(`${zqConfig.apiBaseURL}/api/mobile/v1/links/resolve`, {
    ...baseContext(), ...appBinding(), host: url.host, reference,
    url: url.toString(), parameters: Object.fromEntries(url.searchParams),
  });
  return parseDelivery(response);
}

export async function recoverDeferredLink(handoff: DeferredHandoff): Promise<Delivery> {
  if (!zqConfig.allowedHosts.some(host => host === handoff.host)) {
    throw new Error('Unconfigured deferred-link host');
  }
  const endpoint = Platform.OS === 'ios'
    ? '/api/mobile/v1/deferred/ios/recover'
    : '/api/mobile/v1/deferred/recover';
  const response = await postJSON<unknown>(`${zqConfig.apiBaseURL}${endpoint}`, {
    ...baseContext(), ...appBinding(), token: handoff.token, host: handoff.host,
  });
  return parseDelivery(response);
}

export async function acknowledgeRouteOpened(ack: RouteAcknowledgement, host: string): Promise<void> {
  const url = trustedAcknowledgementURL(ack.endpoint, zqConfig.apiBaseURL);
  await postJSON(url, {receipt: ack.receipt, host, platform, ...appBinding()});
}
